### PR TITLE
NIFI-8761 Enable not setting a value for Escape Character in CSVReade…

### PR DESCRIPTION
**Pull request by timeabarna**
…r controller service
https://issues.apache.org/jira/browse/NIFI-8761

#### Description of PR

Enables to use CSVReader and CSVRecordSetWriter without escape character

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-standard-record-utils/src/main/java/org/apache/nifi/csv/CSVUtils.java [hunk 1]
```diff
@@ -274,7 +275,12 @@ private static CSVFormat buildCustomFormat(final PropertyContext context, final
         final Character quoteChar = getCharUnescaped(context, QUOTE_CHAR, variables);
         format = format.withQuote(quoteChar);
 
-        final Character escapeChar = getCharUnescaped(context, ESCAPE_CHAR, variables);
+        final Character escapeChar;
+        if (context.getProperty(CSVUtils.ESCAPE_CHAR).evaluateAttributeExpressions(variables).getValue().isEmpty()) {
+            escapeChar = null;
+        } else {
+            escapeChar = getCharUnescaped(context, ESCAPE_CHAR, variables);
+        }
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-standard-record-utils/src/main/java/org/apache/nifi/csv/CSVUtils.java:283]: Would the below be easier to read?

```suggestion
        final Character escapeChar = context.getProperty(CSVUtils.ESCAPE_CHAR).evaluateAttributeExpressions(variables).getValue().isEmpty() ? null : getCharUnescaped(context, ESCAPE_CHAR, variables);
```

**Review by pvillard31 — COMMENTED** [review]
Had a very quick look, looks good to me. Made a quick suggestion. Maybe @pgyori might also want to double check this pull request.

**Comment by timeabarna** [comment]
Thanks @pvillard31 for your help, applied your recommendation.

File: nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-standard-record-utils/src/main/java/org/apache/nifi/csv/CSVValidators.java [hunk 2]
```diff
@@ -25,15 +25,18 @@
...
+        illegalChars.add("\n");
+    }
 
-        static {
-            illegalChars.add("\r");
-            illegalChars.add("\n");
-        }
+    public static final Validator SINGLE_CHAR_VALIDATOR = createSingleCharValidator(false);
+
+    public static final Validator EMPTY_AND_SINGLE_CHAR_VALIDATOR = createSingleCharValidator(true);
```
Comment (pgyori) [inline nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-standard-record-utils/src/main/java/org/apache/nifi/csv/CSVValidators.java:37]: I recommend naming it EMPTY_OR_SINGLE_CHAR_VALIDATOR as it contains a logical OR in the implementation. When reading the code of CSVUtils, at `.addValidator(CSVValidators.EMPTY_AND_SINGLE_CHAR_VALIDATOR)` I got a little confused. My first impression was that the property is required to be empty and contain a single character at the same time.

File: nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/test/java/org/apache/nifi/csv/TestCSVValidators.java [hunk 3]
```diff
@@ -33,48 +33,109 @@
...
     @Test
     public void testSingleCharGoodChar() {
-        CSVValidators.SingleCharacterValidator validator = new CSVValidators.SingleCharacterValidator();
+        Validator validator = CSVValidators.SINGLE_CHAR_VALIDATOR;
         ValidationContext mockContext = Mockito.mock(ValidationContext.class);
         ValidationResult result = validator.validate("EscapeChar", "'", mockContext);
         assertTrue(result.isValid());
     }
 
+    /*** Empty And SingleCharValidator **/
```
Comment (pgyori) [inline nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/test/java/org/apache/nifi/csv/TestCSVValidators.java:87]: Being consistent with my other comment, I would replace "And" with "Or".

**Review by pgyori — COMMENTED** [review]
It looks good to me, I only left a minor comment for a suggested name change.
The code looks good.
Tested the functionality and it works as expected.
The documentation is updated correctly and is understandable.
All the newly added code is carefully surrounded with unit tests. Great job!

I found some behavior that is not caused by this modification, but came up for me while testing this. If the input data is:
`"abc\def"`
and the escape character is not specified (null), I found that in the output we get:
`"abc\\def"`
It is present in earlier versions of NiFi, and occurs when we select e.g. RFC 4180 as the CSV Format (in the Reader). I think it might be something that is worth fixing since I suppose it is just a side-effect of having to escape backslashes in strings in Java. I'm not sure though, if it should be handled in this PR or a separate one. @timeabarna , @pvillard31  what do you think?

**Comment by timeabarna** [comment]
@pgyori Iregarding I found some behavior that is not caused by this modification, but came up for me while testing this. If the input data is:
"abc\def"
and the escape character is not specified (null), I found that in the output we get:
"abc\\def"

Is the behaviour is the same if you select null escape character for the Writer as well?
I think it makes sense to add '\' before an '\' if you select '\' as an escape character for the Writer.

**Comment by pgyori** [comment]
Amending my previous comment:
`"abc\def"` turned into `"abc\\def"` because I was using JSONRecordSetWriter, and in JSON it is normal to escape the backslash. When using CSVRecordSetWriter with CSVReader, the "issue" does not occur, so it is fine, no bug to fix there.

**Comment by timeabarna** [comment]
Thanks @pgyori for your help, I've renamed the validator based on your recommendation.

**Review by pgyori — APPROVED** [review]
Thank you @timeabarna !
LGTM +1

**Comment by adenes** [comment]
Thanks @timeabarna for the contribution and @pgyori for the review. Merging to main.

**Pull request merged**